lib/reline/line_editor.rb

```diff
@@ -45,6 +45,8 @@ module CompletionState
   RenderedScreen = Struct.new(:base_y, :lines, :cursor_y, keyword_init: true)
 
   CompletionJourneyState = Struct.new(:line_index, :pre, :target, :post, :list, :pointer)
+  ActionState = Struct.new(:state_type, :state_value)
+  NullActionState = ActionState.new(nil, nil)
```

Comment:
I have used a Struct to make the contents of action_state more understandable, but using a simple Array would also be good.

Reply:
In `ed_search_prev_history` and `ed_search_next_history`, `@prev_action_state == ActionState.new(:search_history, :empty)` is working great :+1:

In general, I think this kind of ideom will appear frequently.
```
foobar = @prev_action_state.value if @prev_action_state&.type == :foobar
```
To make it easy to use, can you add a helper method like this?
```ruby
# get
prev = prev_action_state(:search_history)
# set
set_next_action_state(:search_history, :empty)
```
Using it, we can just use a simple array internally. It will reduce total struct/class count (complexity).
And we don't need to think of NullActionState wrongly modified.

```ruby
state = @prev_action_state
state.value = next_value # bug. this might modify NullActionState
@next_action_state = state
```

Reply:
Looks good! I tried to represent the state in a 2-element array!
19dae5bdffd5e7debad46d0be8f3ee370fe8ec5b